### PR TITLE
Add basic rewrite rule reconstruction

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the module for rewrite reconstruction, and the implementation of basic (rule-independent) rewrite rule reconstruction.

### CODE REVIEW

File: src/smt/proof_post_processor.cpp [hunk 1]
```diff
@@ -842,15 +843,16 @@ Node ProofPostprocessCallback::expandMacros(PfRule id,
...
-    // first, check that maybe its just an evaluation step
-    ProofChecker* pc = pnm->getChecker();
-    Node ceval = pc->checkDebug(
-        PfRule::EVALUATE, {}, {eq[0]}, Node::null(), "smt-proof-pp-debug");
-    if (!ceval.isNull() && ceval == eq)
+    TheoryId tid = THEORY_BUILTIN;
+    builtin::BuiltinProofRuleChecker::getTheoryId(args[1], tid);
+    MethodId mid = MethodId::RW_REWRITE;
+    getMethodId(args[2], mid);
+    uint32_t recLimit = options().proof.proofRewriteRconsRecLimit;
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:850]: The type in the option declaration is `int64_t`. Shouldn't these match?

File: src/rewriter/rewrite_db_proof_cons.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
```
Comment (HanielB) [inline src/rewriter/rewrite_db_proof_cons.h:7]:
```suggestion
 * Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
```
This file has been waiting for a while. :)

File: src/rewriter/rewrite_db_proof_cons.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#include "rewriter/rewrites.h"
+#include "smt/env_obj.h"
+#include "theory/evaluator.h"
+#include "util/statistics_stats.h"
+
+namespace cvc5::internal {
+namespace rewriter {
+
+/**
+ * This class is used to reconstruct proofs of theory rewrites.
```
Comment (HanielB) [inline src/rewriter/rewrite_db_proof_cons.h:39]: It'd be nice to add a reference to the FMCAD paper.

File: src/rewriter/rewrite_db_proof_cons.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  bool prove(CDProof* cdp,
+             Node a,
+             Node b,
+             theory::TheoryId tid,
+             MethodId mid,
+             uint32_t recLimit);
+
+ private:
+  /** Basic utility */
```
Comment (HanielB) [inline src/rewriter/rewrite_db_proof_cons.h:57]: clarify

File: src/rewriter/basic_rewrite_rcons.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      Trace("trewrite-rcons") << "...BETA_REDUCE" << std::endl;
+      return true;
+    }
+  }
+  Trace("trewrite-rcons") << "...(fail)" << std::endl;
+  return false;
+}
+
+bool BasicRewriteRCons::tryRule(CDProof* cdp,
+                                 Node eq,
```
Comment (HanielB) [inline src/rewriter/basic_rewrite_rcons.cpp:66]: clang-format

Comment (HanielB) [reply src/rewriter/basic_rewrite_rcons.cpp:66]: there are other instances also

**Review by HanielB — APPROVED** [review]
LGTM, just some minor comments

File: src/rewriter/rewrite_db_proof_cons.h (hunk 2 above)
Comment (ajreynol) [reply src/rewriter/rewrite_db_proof_cons.h:7]: Yes it has, :-)

File: src/smt/proof_post_processor.cpp (hunk 1 above)
Comment (ajreynol) [reply src/smt/proof_post_processor.cpp:890]: Yes, this must have gotten out of sync.

**Pull request merged**